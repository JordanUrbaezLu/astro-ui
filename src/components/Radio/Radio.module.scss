@use "../../utility/globals";

.container {
  align-items: center;
  display: flex;
  gap: 4px;
  width: fit-content;
}

.radioIcon {
  background-color: transparent !important;
  color: globals.$secondary-text-color !important;

  &.isSelected {
    color: globals.$primary-background-color !important;
  }

  &:hover,
  &:focus {
    background-color: globals.$secondary-background-color-hover-and-active !important;
  }

  &:active {
    background-color: globals.$secondary-background-color-active !important;
  }
}

.selected {
  color: globals.$primary-background-color !important;
}
