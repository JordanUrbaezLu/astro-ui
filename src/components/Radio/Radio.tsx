import * as React from "react";
import classNames from "classnames";
import styles from "./Radio.module.scss";
import {
  MdRadioButtonUnchecked,
  MdRadioButtonChecked,
} from "react-icons/md";
import { IconButton } from "../IconButton";
import { Typography } from "../Typography/Typography";

export interface RadioProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The label for the Radio
   */
  children: React.ReactNode;
  /**
   * If the Radio is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * If the Radio is selected
   *
   * @default false
   */
  isSelected?: boolean;
  /**
   * The callback fired when requesting to change the Radio state
   */
  onChange: () => void;
}

/**
 * @public
 *
 * @description
 *
 * The Radio allows a user to select an item.
 */
export const Radio: React.FC<RadioProps> = (props) => {
  const {
    children,
    className,
    disabled = false,
    isSelected = false,
    onChange,
    ...rest
  } = props;

  return (
    <div className={styles.container}>
      <IconButton
        className={classNames(
          className,
          isSelected && styles.isSelected
        )}
        content={
          isSelected ? (
            <MdRadioButtonChecked />
          ) : (
            <MdRadioButtonUnchecked />
          )
        }
        disabled={disabled}
        onClick={onChange}
        role="radio"
        variant="secondary"
        {...rest}
      />
      <Typography
        className={classNames(
          disabled && styles.disabled,
          isSelected && styles.selected
        )}
        variant="secondary"
      >
        {children}
      </Typography>
    </div>
  );
};

Radio.displayName = "Radio";
