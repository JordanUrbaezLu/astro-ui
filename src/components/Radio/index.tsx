/* istanbul ignore file */
export * from "./Radio";
