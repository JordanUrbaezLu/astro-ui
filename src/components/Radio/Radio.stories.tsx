import * as React from "react";
import { Radio } from "./Radio";

export default {
  component: Radio,
  title: "Components/Inputs/Radio",
};

export const Default = () => {
  const [isSelected, setIsSelected] = React.useState<boolean>(false);

  return (
    <Radio
      isSelected={isSelected}
      onChange={() => setIsSelected(!isSelected)}
    >
      Default
    </Radio>
  );
};

export const Disabled = () => {
  const [isSelected, setIsSelected] = React.useState<boolean>(false);

  return (
    <Radio
      disabled
      isSelected={isSelected}
      onChange={() => setIsSelected(!isSelected)}
    >
      Disabled
    </Radio>
  );
};

export const DisabledAndChecked = () => {
  const [isSelected, setIsSelected] = React.useState<boolean>(true);

  return (
    <Radio
      disabled
      isSelected={isSelected}
      onChange={() => setIsSelected(!isSelected)}
    >
      Disabled
    </Radio>
  );
};

export const Selected = () => {
  const [isSelected, setIsSelected] = React.useState<boolean>(true);

  return (
    <Radio
      isSelected={isSelected}
      onChange={() => setIsSelected(!isSelected)}
    >
      Selected
    </Radio>
  );
};
