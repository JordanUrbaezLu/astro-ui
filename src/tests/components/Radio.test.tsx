import { render, screen } from "@testing-library/react";
import { Radio } from "../../components/Radio/Radio";
import { axe } from "jest-axe";

describe("Switch", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Radio isSelected onChange={jest.fn()}>
        Radio
      </Radio>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Radio isSelected onChange={jest.fn()}>
        Radio
      </Radio>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render label correctly.", () => {
    render(
      <Radio isSelected onChange={jest.fn()}>
        Radio
      </Radio>
    );
    expect(screen.getByText("Radio")).toBeInTheDocument();
  });

  test("Should render unchecked correctly.", () => {
    render(<Radio onChange={jest.fn()}>Radio</Radio>);
    expect(screen.getByText("Radio")).not.toHaveClass("selected");
  });

  test("Should render disabled correctly.", () => {
    render(
      <Radio disabled onChange={jest.fn()}>
        Radio
      </Radio>
    );
    expect(screen.getByText("Radio")).toHaveClass("disabled");
  });
});
