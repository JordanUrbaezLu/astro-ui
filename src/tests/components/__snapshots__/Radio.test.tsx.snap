// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Should render component correctly. 1`] = `
<div
  class="container"
>
  <button
    aria-label="Icon Button"
    class="isSelected iconButton medium secondary"
    role="radio"
  >
    <div
      class="container"
      role="presentation"
    />
    <svg
      fill="currentColor"
      height="22"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="22"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
      />
    </svg>
  </button>
  <div
    class="selected typography typographyh2 secondary normal"
  >
    Radio
  </div>
</div>
`;
